composer/loggers/mosaicml_logger.py

```diff
@@ -96,6 +102,90 @@ def log_hyperparameters(self, hyperparameters: Dict[str, Any]):
     def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None) -> None:
         self._log_metadata(metrics)
 
+    def log_analytics(self,) -> None:
+        if self.analytics_data is None:
+            return
+
+        # Fetch / cast metrics that we want to log from self.analytics_data
+        autoresume: bool = self.analytics_data['autoresume']
+        trainer_state: State = self.analytics_data['state']
+        save_interval: Union[str, int, Time, Callable[[State, Event], bool]] = self.analytics_data['save_interval']
+        loggers: List[LoggerDestination] = self.analytics_data['loggers']
+        load_path: Union[str, None] = self.analytics_data['load_path']
+        save_folder: Union[str, None] = self.analytics_data['save_folder']
+
+        metrics: Dict[str, Any] = {'composer/autoresume': autoresume, 'composer/precision': trainer_state.precision}
+
+        train_dataloader = trainer_state.train_dataloader
+        if train_dataloader is not None and isinstance(train_dataloader, torch.utils.data.DataLoader):
+            metrics['composer/train_loader_workers'] = train_dataloader.num_workers
+
+        metrics['composer/eval_loaders'] = []
+        for evaluator in trainer_state.evaluators:
+            dataloader = evaluator.dataloader.dataloader
+            if isinstance(dataloader, torch.utils.data.DataLoader):
+                metrics['composer/eval_loaders'].append(
+                    json.dumps({
+                        'label': evaluator.label,
+                        'num_workers': dataloader.num_workers,
+                    }),
+                )
+
+        metrics['composer/optimizers'] = [
+            json.dumps(optimizer.state_dict(), sort_keys=True) for optimizer in trainer_state.optimizers
```

Comment:
I think this optimizer state dict can be very large -- see [here](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.state_dict.html#torch.optim.Optimizer.state_dict) for reference. Calling `optimizer.state_dict()` like this probably gives us the per-parameter state of the optimizer, which will be a ton of data. Has this logging been tested with a decent-sized model, and do we save what you expect? Mind adding a section about testing to the PR description?

Reply:
<img width="1324" alt="Screenshot 2024-03-19 at 5 28 05 PM" src="https://github.com/mosaicml/composer/assets/141283633/4bb674c3-c3d1-4758-b6b1-67cbaa94b309">

here's the logs after running the [quickstart example in the README](https://github.com/mosaicml/composer/tree/angel/add-data-to-metadata-for-analytics?tab=readme-ov-file#quick-start). 

looks like the state dict will get pretty huge -- are there any values that you think will be useful to capture from here? don't want to be logging this unless it's useful.

for `callbacks` maybe we can try to consolidate this too? would it be suitable to just have a list of the `repr` values?

@snarayan21

Reply:
This is small for SGD and MASSIVE for real optimizers (eg adamw). I would just log the name of the optimizer and call it a day